torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -813,38 +813,52 @@ at::IntArrayRef strides_or_error(const Tensor & input, c10::string_view const &
       "Please either use a strided tensor or set requires_grad=False for '",
       input_name, "'");
     if (input.is_mkldnn()) return IntArrayRef({});
+    if (input.is_sparse_csr()) return IntArrayRef({});
     return input.strides();
   } else {
     return IntArrayRef({});
   }
 }
 
-Tensor mm_mat1_backward(const Tensor & grad, const Tensor & mat2, at::IntArrayRef mat1_sizes, at::IntArrayRef mat1_strides, const Scalar & alpha) {
-  // if input was column-major, return grad as column-order for efficiency
-  if (mat1_strides[0] == 1 && mat1_strides[1] == mat1_sizes[0]) {
-    return maybe_multiply(mat2.conj().mm(grad.t()).t(), alpha.conj());
-  } else {
-    return maybe_multiply(grad.mm(mat2.t().conj()), alpha.conj());
-  }
-}
-
-Tensor mm_mat2_backward(const Tensor & grad, const Tensor & mat1, IntArrayRef sizes, IntArrayRef strides, const Scalar & alpha) {
-  // if input was column-major, return grad as column-order for efficiency
-  if (strides[0] == 1 && strides[1] == sizes[0]) {
-    if (mat1.is_sparse()) {
-      // Since mm(dense, sparse) doesn't exist,
-      // pass a transposed output matrix to the underlying "addmm"
-      // function directly.
-      int64_t out_rows = mat1.size(1);
-      int64_t out_cols = grad.size(1);
-      Tensor t = at::zeros({}, grad.options()).expand({out_rows, out_cols}, true);
-      Tensor r = at::empty({out_cols, out_rows}, grad.options()).t();
-      at::addmm_out(r, t, mat1.t(), grad, alpha, 1);
-      return r;
+Tensor mm_mat1_backward(const Tensor& grad, const Tensor& mat2, at::IntArrayRef mat1_sizes, at::IntArrayRef mat1_strides, c10::Layout mat1_layout, const Scalar& alpha) {
+  if (grad.layout() == c10::kStrided && mat2.layout() == c10::kStrided && mat1_layout == c10::kStrided) {
+    // if input was column-major, return grad as column-order for efficiency
+    if (mat1_strides[0] == 1 && mat1_strides[1] == mat1_sizes[0]) {
+      return maybe_multiply(mat2.conj().mm(grad.t()).t(), alpha.conj());
     }
-    return maybe_multiply(grad.t().mm(mat1.conj()).t(), alpha.conj());
-  } else {
-    return maybe_multiply(mat1.t().conj().mm(grad), alpha.conj());
+  }
+
+  auto result = maybe_multiply(grad.mm(mat2.t().conj()), alpha.conj());
+
+  switch (mat1_layout) {
```

Comment:
Why no sparse on this one while it is there for the mat2 backward?

Reply:
sparse_coo with requires_grad=True is intercepted by `strides_or_error` function and raises an error. I thought it would be better to handle the COO separately later.

Reply:
Removed handling of layout per our off-GH discussion.